### PR TITLE
[SDK] Restore Recordable API compatibility with versions < 1.14.0

### PR DESCRIPTION
**Pull request by dbolduc**
Issue:

`Recordable::SetTraceFlags(...)` was added in v1.14.0. This was a breaking change. Consumers of this library would have to define an override for this method. Also, the latest version of opentelemetry-cpp would not be API compatible with older versions of consumers (which could not know about this new, pure virtual method).

https://github.com/open-telemetry/opentelemetry-cpp/pull/2488#discussion_r1493959370

Solution:

Provides a default implementation for `Recordable::SetTraceFlags(...)`.

The default implementation is sufficient for W3C Trace Context version 1, which is actually used.

To later support W3C Trace Context version 2, exporters are encouraged to capture trace flags, by providing an implementation in Recordable sub classes.

Limitation:

Note that this PR does not restore ABI compatibility for older versions.

The ABI change is required to support opentelemetry-proto 1.1.0, used by the OTLP exporter.

### CODE REVIEW

**Comment by lalitb** [comment]
Agree with the change. This got overlooked in the review.

**Comment by marcalff** [comment]
@dbolduc 

Thanks for the details.

This is what I suggest then:

1) Merge #2549 to fix the release notes, it is better anyway.

2) Merge this PR to provide a default no-op instrumentation.

It is correct that some exporters will never care about trace flags (like zipkin).

My first concern is with:

> B. provide a default implementation that might go unnoticed by exporters that should use the trace flags

but it can be mitigated by other means (documentation) instead of forcing a decision with a build break (pure virtual method).

Ok to have 1.14.1 as well.

My second concern is with binary compatibility.

With of without a no op implementation, mixing binaries pre and post 1.14.0 will lead to crashes, so the claim that "Recordable is backward compatible" should be clarified and put into context, because there is still an ABI change that can not be ignored.

**Comment by dbolduc** [comment]
> This is what I suggest then...

Thanks for working with me on this.

> The claim that "Recordable is backward compatible" should be clarified and put into context

I updated the PR title and description to try to capture the nuance. (sorry, I only ever think about API compatibility). Feel free to make edits.

File: sdk/include/opentelemetry/sdk/trace/recordable.h [hunk 1]
```diff
@@ -139,7 +139,7 @@ class Recordable
    * Set the trace flags of the span.
    * @param flags the flags to set
    */
-  virtual void SetTraceFlags(opentelemetry::trace::TraceFlags flags) noexcept = 0;
+  virtual void SetTraceFlags(opentelemetry::trace::TraceFlags flags) noexcept { (void)flags; }
```
[the remark this answers is not in the document]
Comment (marcalff) [reply sdk/include/opentelemetry/sdk/trace/recordable.h]: Fixed by commenting the flags parameter name instead.

**Comment by marcalff** [comment]
@dbolduc 

Done with edits, please confirm you are ok with the rewording.

**Comment by dbolduc** [comment]
LGTM, thanks!

**Review by marcalff — APPROVED** [review]

**Pull request merged**